### PR TITLE
Hyperparameter Search: Run learning multiple times for each configuration

### PR DESCRIPTION
**Pull request by luator**
## Description

To reduce the noise in the estimated reward for a given configuration, run the learning multiple times in each job and use mean reward as metric for the optimisation.

This is implemented using `cluster.exit_for_resume()`, i.e. for each iteration of learning the cluster job is restarted.  This way the
duration of the single cluster jobs does not increase and it also makes it easier to handle the restarts in case of failure more properly. The "restarts" log file is extended for this to not only include the number of failed runs but also the rewards of the finished runs and a "finished runs" counter.

Further changes:

- Use ".json" for restarts log. Since the config file is explicitly
  passed to the hysr scripts now, there is no reason to not use ".json"
  here.
- Make the max. number of retries in case of failures configurable.
- Disable warnings caused by tensorflow (to reduce the noise in the stderr output)

I'm planning to refactor the run_learning.py script but thought it might be good to already merge the current state, so the feature can already be used.

**Note:** `cluster.exit_for_resume()` seems not to work when running locally (i.e. not on the cluster), so unfortunately it is not possible to test this locally anymore unless that issue gets fixed.

## How I Tested

By running it on the cluster with very short training duration.

## I fulfilled the following requirements

[//]: # "Please make sure you followed these steps before requesting a review."
[//]: # "Check the boxes in the list below, when done."

- [x] All new code is formatted according to our style guide (for C++ run clang-format, for Python, run flake8 and fix all warnings).
- [x] All new functions/classes are documented and existing documentation is updated according to changes.
- [x] No commented code from testing/debugging is kept (unless there is a good reason to keep it).

### CODE REVIEW

File: hp_optim/run_learning.py [hunk 1]
```diff
@@ -141,15 +148,75 @@ def read_reward_from_log(log_dir: pathlib.PurePath):
     return eprewmean
 
 
+# TODO maybe better to implement this in a class
+def run_learning(config_file, env, proc_backend):  # TODO type hints
```
Comment (vincentberenz) [inline hp_optim/run_learning.py:152]: TODO: why not now ? :)

**Review by vincentberenz — APPROVED** [review]

File: hp_optim/run_learning.py (hunk 1 above)
Comment (luator) [reply hp_optim/run_learning.py:152]: Indeed, I missed that TODO.  Thanks for pointing it out!

Comment (luator) [reply hp_optim/run_learning.py]: Added type hints and docstring.
I plan to address the other TODO in a separate PR (with some bigger refactoring of the script) to not delay merging this "multiple runs" feature.
